In [1]:
source('../src/load_data.r')
source('../src/multiplot.r')

In [2]:
head(categoric_df,3)
t(dim(categoric_df))
head(numeric_updated_df,3)
t(dim(numeric_updated_df))

MSSubClass MSZoning LotShape LandContour Utilities LotConfig LandSlope
1 60         RL       Reg      Lvl         AllPub    Inside    Gtl      
2 20         RL       Reg      Lvl         AllPub    FR2       Gtl      
3 60         RL       IR1      Lvl         AllPub    Inside    Gtl      
  Neighborhood Condition1 Condition2 ⋯ GarageQual GarageCond PavedDrive PoolQC 
1 CollgCr      Norm       Norm       ⋯ TA         TA         Y          without
2 Veenker      Feedr      Norm       ⋯ TA         TA         Y          without
3 CollgCr      Norm       Norm       ⋯ TA         TA         Y          without
  Fence   MiscFeature MoSold YrSold SaleType SaleCondition
1 without without     2      2008   WD       Normal       
2 without without     5      2007   WD       Normal       
3 without without     9      2008   WD       Normal

[,1] [,2]
[1,] 1451 57

X BsmtFinSF2 LowQualFinSF EnclosedPorch ThreeSsnPorch ScreenPorch PoolArea
1 1 FALSE      FALSE        FALSE         FALSE         FALSE       FALSE   
2 2 FALSE      FALSE        FALSE         FALSE         FALSE       FALSE   
3 3 FALSE      FALSE        FALSE         FALSE         FALSE       FALSE   
  MiscVal MasVnrArea SecondFlrSF ⋯ YearBuilt YearRemodAdd BsmtFinSF1 BsmtUnfSF 
1 FALSE   1st_bin    2nd_bin     ⋯ 1.0509849  0.8820760   0.7205018  0.08917398
2 FALSE   0          0           ⋯ 0.1692304 -0.4188065   0.7552734  0.20492237
3 FALSE   1st_bin    2nd_bin     ⋯ 0.9860786  0.8342087   0.6806608  0.28181796
  TotalBsmtSF FirstFlrSF GrLivArea  GarageYrBlt GarageArea SalePrice
1 0.1084214   -0.8026317  0.5315369  1.02265460 0.2771098  0.5639683
2 0.2264173    0.4214040 -0.3785471 -0.09204807 0.2390487  0.2168273
3 0.1303386   -0.5752722  0.6618057  0.94060083 0.2997008  0.7378659

[,1] [,2]
[1,] 1451 24

In [3]:
ames_df <- data.frame(categoric_df, numeric_updated_df)
ames_df$X <- NULL
head(ames_df)
t(dim(ames_df))

MSSubClass MSZoning LotShape LandContour Utilities LotConfig LandSlope
1 60         RL       Reg      Lvl         AllPub    Inside    Gtl      
2 20         RL       Reg      Lvl         AllPub    FR2       Gtl      
3 60         RL       IR1      Lvl         AllPub    Inside    Gtl      
4 70         RL       IR1      Lvl         AllPub    Corner    Gtl      
5 60         RL       IR1      Lvl         AllPub    FR2       Gtl      
6 50         RL       IR1      Lvl         AllPub    Inside    Gtl      
  Neighborhood Condition1 Condition2 ⋯ YearBuilt  YearRemodAdd BsmtFinSF1
1 CollgCr      Norm       Norm       ⋯  1.0509849  0.8820760   0.7205018 
2 Veenker      Feedr      Norm       ⋯  0.1692304 -0.4188065   0.7552734 
3 CollgCr      Norm       Norm       ⋯  0.9860786  0.8342087   0.6806608 
4 Crawfor      Norm       Norm       ⋯ -1.8680714 -0.7103048   0.5941378 
5 NoRidge      Norm       Norm       ⋯  0.9536011  0.7384023   0.7125017 
6 Mitchel      Norm       Norm       ⋯  0.7258029  0.4984665   0.7243605 
  BsmtUnfSF   TotalBsmtSF FirstFlrSF GrLivArea  GarageYrBlt GarageArea
1  0.08917398 0.10842138  -0.8026317  0.5315369  1.02265460 0.2771098 
2  0.20492237 0.22641726   0.4214040 -0.3785471 -0.09204807 0.2390487 
3  0.28181796 0.13033862  -0.5752722  0.6618057  0.94060083 0.2997008 
4  0.32144248 0.07065950  -0.4377884  0.5437750  0.81736628 0.3115320 
5  0.30382402 0.19684307   0.1146148  1.2836237  0.89954319 0.3689426 
6 -0.06527209 0.08633152  -1.0317819 -0.1501048  0.61156359 0.2483025 
  SalePrice 
1  0.5639683
2  0.2168273
3  0.7378659
4 -0.4330156
5  1.0183392
6 -0.3799439

[,1] [,2]
[1,] 1451 80

In [4]:
str(ames_df)

'data.frame':	1451 obs. of  80 variables:
 $ MSSubClass   : Factor w/ 15 levels "20","30","40",..: 6 1 6 7 6 5 1 6 5 15 ...
 $ MSZoning     : Factor w/ 5 levels "C (all)","FV",..: 4 4 4 4 4 4 4 4 5 4 ...
 $ LotShape     : Factor w/ 4 levels "IR1","IR2","IR3",..: 4 4 1 1 1 1 4 1 4 4 ...
 $ LandContour  : Factor w/ 4 levels "Bnk","HLS","Low",..: 4 4 4 4 4 4 4 4 4 4 ...
 $ Utilities    : Factor w/ 2 levels "AllPub","NoSeWa": 1 1 1 1 1 1 1 1 1 1 ...
 $ LotConfig    : Factor w/ 5 levels "Corner","CulDSac",..: 5 3 5 1 3 5 5 1 5 1 ...
 $ LandSlope    : Factor w/ 3 levels "Gtl","Mod","Sev": 1 1 1 1 1 1 1 1 1 1 ...
 $ Neighborhood : Factor w/ 25 levels "Blmngtn","Blueste",..: 6 25 6 7 14 12 21 17 18 4 ...
 $ Condition1   : Factor w/ 9 levels "Artery","Feedr",..: 3 2 3 3 3 3 3 5 1 1 ...
 $ Condition2   : Factor w/ 8 levels "Artery","Feedr",..: 3 3 3 3 3 3 3 3 3 1 ...
 $ BldgType     : Factor w/ 5 levels "1Fam","2fmCon",..: 1 1 1 1 1 1 1 1 1 2 ...
 $ HouseStyle   : Factor w/ 8 levels "1.5Fin","1.

### One Hot Encoding Categorical Variables

In [5]:
library(caret)

Loading required package: lattice


In [6]:
new_categorical_df <- Filter(is.factor, ames_df)
t(colnames(new_categorical_df))
length(colnames(new_categorical_df))

[,1]       [,2]     [,3]     [,4]        [,5]      [,6]      [,7]     
[1,] MSSubClass MSZoning LotShape LandContour Utilities LotConfig LandSlope
     [,8]         [,9]       [,10]      [,11] [,12] [,13]       [,14]  [,15] 
[1,] Neighborhood Condition1 Condition2 ⋯     Fence MiscFeature MoSold YrSold
     [,16]    [,17]         [,18]      [,19]       [,20]      [,21]      
[1,] SaleType SaleCondition MasVnrArea SecondFlrSF WoodDeckSF OpenPorchSF

[1] 61

In [7]:
dummy_df <- dummyVars(" ~ .", data=new_categorical_df)
dummy_df

Dummy Variable Object

Formula: ~.
<environment: 0xaa10a50>
61 variables, 61 factors
Variables and levels will be separated by '.'
A less than full rank encoding is used

In [8]:
ohe_categorical_df <- data.frame(predict(dummy_df, newdata = new_categorical_df))
head(ohe_categorical_df)
t(dim(ohe_categorical_df))

MSSubClass.20 MSSubClass.30 MSSubClass.40 MSSubClass.45 MSSubClass.50
1 0             0             0             0             0            
2 1             0             0             0             0            
3 0             0             0             0             0            
4 0             0             0             0             0            
5 0             0             0             0             0            
6 0             0             0             0             1            
  MSSubClass.60 MSSubClass.70 MSSubClass.75 MSSubClass.80 MSSubClass.85 ⋯
1 1             0             0             0             0             ⋯
2 0             0             0             0             0             ⋯
3 1             0             0             0             0             ⋯
4 0             1             0             0             0             ⋯
5 1             0             0             0             0             ⋯
6 0             0             0             0             0             ⋯
  WoodDeckSF.0 WoodDeckSF.1st_bin WoodDeckSF.2nd_bin WoodDeckSF.3rd_bin
1 1            0                  0                  0                 
2 0            0                  1                  0                 
3 1            0                  0                  0                 
4 1            0                  0                  0                 
5 0            1                  0                  0                 
6 0            1                  0                  0                 
  WoodDeckSF.4th_bin OpenPorchSF.0 OpenPorchSF.1st_bin OpenPorchSF.2nd_bin
1 0                  0             1                   0                  
2 0                  1             0                   0                  
3 0                  0             1                   0                  
4 0                  0             1                   0                  
5 0                  0             1                   0                  
6 0                  0             1                   0                  
  OpenPorchSF.3rd_bin OpenPorchSF.4th_bin
1 0                   0                  
2 0                   0                  
3 0                   0                  
4 0                   0                  
5 0                   0                  
6 0                   0

[,1] [,2]
[1,] 1451 400

In [9]:
scaled_ohe_categorical_df <- scale(ohe_categorical_df)
head(scaled_ohe_categorical_df)

MSSubClass.20 MSSubClass.30 MSSubClass.40 MSSubClass.45 MSSubClass.50
1 -0.760586     -0.2233679    -0.05255887   -0.09128733   -0.3318133   
2  1.313870     -0.2233679    -0.05255887   -0.09128733   -0.3318133   
3 -0.760586     -0.2233679    -0.05255887   -0.09128733   -0.3318133   
4 -0.760586     -0.2233679    -0.05255887   -0.09128733   -0.3318133   
5 -0.760586     -0.2233679    -0.05255887   -0.09128733   -0.3318133   
6 -0.760586     -0.2233679    -0.05255887   -0.09128733    3.0116663   
  MSSubClass.60 MSSubClass.70 MSSubClass.75 MSSubClass.80 MSSubClass.85 ⋯
1  1.9746741    -0.2076167    -0.1055563    -0.2021419    -0.1181804    ⋯
2 -0.5060637    -0.2076167    -0.1055563    -0.2021419    -0.1181804    ⋯
3  1.9746741    -0.2076167    -0.1055563    -0.2021419    -0.1181804    ⋯
4 -0.5060637     4.8132480    -0.1055563    -0.2021419    -0.1181804    ⋯
5  1.9746741    -0.2076167    -0.1055563    -0.2021419    -0.1181804    ⋯
6 -0.5060637    -0.2076167    -0.1055563    -0.2021419    -0.1181804    ⋯
  WoodDeckSF.0 WoodDeckSF.1st_bin WoodDeckSF.2nd_bin WoodDeckSF.3rd_bin
1  0.9571546   -0.7094222         -0.3761957         -0.1268674        
2 -1.0440433   -0.7094222          2.6563593         -0.1268674        
3  0.9571546   -0.7094222         -0.3761957         -0.1268674        
4  0.9571546   -0.7094222         -0.3761957         -0.1268674        
5 -1.0440433    1.4086264         -0.3761957         -0.1268674        
6 -1.0440433    1.4086264         -0.3761957         -0.1268674        
  WoodDeckSF.4th_bin OpenPorchSF.0 OpenPorchSF.1st_bin OpenPorchSF.2nd_bin
1 -0.05878292        -0.9055449     1.0822883          -0.2891139         
2 -0.05878292         1.1035464    -0.9233315          -0.2891139         
3 -0.05878292        -0.9055449     1.0822883          -0.2891139         
4 -0.05878292        -0.9055449     1.0822883          -0.2891139         
5 -0.05878292        -0.9055449     1.0822883          -0.2891139         
6 -0.05878292        -0.9055449     1.0822883          -0.2891139         
  OpenPorchSF.3rd_bin OpenPorchSF.4th_bin
1 -0.0950479          -0.05255887        
2 -0.0950479          -0.05255887        
3 -0.0950479          -0.05255887        
4 -0.0950479          -0.05255887        
5 -0.0950479          -0.05255887        
6 -0.0950479          -0.05255887

### Turn Logical Variables into Integers

In [10]:
bools_df <- Filter(is.logical, ames_df)
t(colnames(bools_df))
t(dim(bools_df))

[,1]       [,2]         [,3]          [,4]          [,5]        [,6]    
[1,] BsmtFinSF2 LowQualFinSF EnclosedPorch ThreeSsnPorch ScreenPorch PoolArea
     [,7]   
[1,] MiscVal

[,1] [,2]
[1,] 1451 7

In [12]:
int_bool_df <- as.data.frame(lapply(bools_df, as.integer))
scaled_ohebool_df <- scale(int_bool_df)
head(scaled_ohebool_df)

BsmtFinSF2 LowQualFinSF EnclosedPorch ThreeSsnPorch ScreenPorch
[1,] -0.3605172 -0.1350298   -0.4077794    -0.1296415    -0.2946719 
[2,] -0.3605172 -0.1350298   -0.4077794    -0.1296415    -0.2946719 
[3,] -0.3605172 -0.1350298   -0.4077794    -0.1296415    -0.2946719 
[4,] -0.3605172 -0.1350298    2.4506163    -0.1296415    -0.2946719 
[5,] -0.3605172 -0.1350298   -0.4077794    -0.1296415    -0.2946719 
[6,] -0.3605172 -0.1350298   -0.4077794     7.7082654    -0.2946719 
     PoolArea    MiscVal   
[1,] -0.06960104 -0.1927272
[2,] -0.06960104 -0.1927272
[3,] -0.06960104 -0.1927272
[4,] -0.06960104 -0.1927272
[5,] -0.06960104 -0.1927272
[6,] -0.06960104  5.1851041

In [13]:
numeric_df <- Filter(is.numeric, ames_df)
t(dim(numeric_df))

[,1] [,2]
[1,] 1451 12

In [14]:
final_ames_df <- data.frame(scaled_ohe_categorical_df, scaled_ohebool_df, numeric_df)
unscaled_ames_df <- data.frame(ohe_categorical_df, int_bool_df, numeric_df)

In [16]:
t(dim(final_ames_df))
t(dim(unscaled_ames_df))

[,1] [,2]
[1,] 1451 419

[,1] [,2]
[1,] 1451 419

In [17]:
### Checking data types for final_ames_df
any_factors <- Filter(is.factor, final_ames_df)
t(dim(any_factors))

any_bools <- Filter(is.logical, final_ames_df)
t(dim(any_bools))

any_numeric <- Filter(is.numeric, final_ames_df)
t(dim(any_numeric))

[,1] [,2]
[1,] 1451 0

[,1] [,2]
[1,] 1451 0

[,1] [,2]
[1,] 1451 419

In [18]:
### Checking data types for unscaled_ames_df
any_factors <- Filter(is.factor, unscaled_ames_df)
t(dim(any_factors))

any_bools <- Filter(is.logical, unscaled_ames_df)
t(dim(any_bools))

any_numeric <- Filter(is.numeric, unscaled_ames_df)
t(dim(any_numeric))

[,1] [,2]
[1,] 1451 0

[,1] [,2]
[1,] 1451 0

[,1] [,2]
[1,] 1451 419

In [19]:
#After OHE categorical variables that originally contained "NA" values, a column with NAs appears in the OHE.

final_ames_df$Alley. <- NULL
final_ames_df$MasVnrType. <- NULL
final_ames_df$BsmtQual. <- NULL
final_ames_df$BsmtCond. <- NULL
final_ames_df$BsmtExposure. <- NULL
final_ames_df$BsmtFinType1. <- NULL
final_ames_df$BsmtFinType2. <- NULL
final_ames_df$Electrical. <- NULL
final_ames_df$FireplaceQu. <- NULL
final_ames_df$GarageType. <- NULL
final_ames_df$GarageFinish. <- NULL
final_ames_df$GarageQual. <- NULL
final_ames_df$GarageCond. <- NULL
final_ames_df$PoolQC. <- NULL
final_ames_df$Fence. <- NULL
final_ames_df$MiscFeature. <- NULL


unscaled_ames_df$Alley. <- NULL
unscaled_ames_df$MasVnrType. <- NULL
unscaled_ames_df$BsmtQual. <- NULL
unscaled_ames_df$BsmtCond. <- NULL
unscaled_ames_df$BsmtExposure. <- NULL
unscaled_ames_df$BsmtFinType1. <- NULL
unscaled_ames_df$BsmtFinType2. <- NULL
unscaled_ames_df$Electrical. <- NULL
unscaled_ames_df$FireplaceQu. <- NULL
unscaled_ames_df$GarageType. <- NULL
unscaled_ames_df$GarageFinish. <- NULL
unscaled_ames_df$GarageQual. <- NULL
unscaled_ames_df$GarageCond. <- NULL
unscaled_ames_df$PoolQC. <- NULL
unscaled_ames_df$Fence. <- NULL
unscaled_ames_df$MiscFeature. <- NULL

In [20]:
t(dim(final_ames_df))
t(dim(unscaled_ames_df))

[,1] [,2]
[1,] 1451 403

[,1] [,2]
[1,] 1451 403

Our final dataframe consists of 1451 rows and 402 columns, which are now all numeric values.

In [21]:
write.csv(final_ames_df, 'final_ames_df.csv')

In [22]:
write.csv(unscaled_ames_df, 'unscaled_ames_df.csv')